src/Stream.sol

```diff
@@ -228,110 +238,93 @@ contract Stream is IStream, Clone {
+        // When the stream is under-funded, it should keep its current balance
+        // When it's sufficiently-funded, it should keep the full balance committed to recipient
+        // i.e. `remainingBalance` or `recipientCancelBalance`
+        uint256 requiredBalanceAfter =
+            Math.min(tokenBalance(), Math.max(remainingBalance, recipientCancelBalance));
 
         IERC20(tokenAddress).safeTransfer(msg.sender, amount);
-    }
 
-    /**
-     * ░░░░░░░░░░░░░░░░░░░░░░░░░░░░░
-     *   VIEW FUNCTIONS
-     * ░░░░░░░░░░░░░░░░░░░░░░░░░░░░░
-     */
+        if (tokenBalance() < requiredBalanceAfter) revert RescueTokenAmountExceedsExcessBalance();
 
-    /**
-     * @notice Returns the available funds to withdraw.
-     * @param who The address for which to query the balance.
-     * @return uint256 The total funds allocated to `who` as uint256.
-     */
-    function balanceOf(address who) public view returns (uint256) {
-        uint256 recipientBalance = _recipientBalance();
-
-        if (who == recipient()) return recipientBalance;
-        if (who == payer()) {
-            // This is safe because it should always be the case that:
-            // remainingBalance >= recipientBalance.
-            unchecked {
-                return remainingBalance - recipientBalance;
-            }
-        }
-        return 0;
-    }
-
-    /**
-     * @notice Returns the time elapsed in this stream, or zero if it hasn't started yet.
-     */
-    function elapsedTime() public view returns (uint256) {
-        uint256 startTime_ = startTime();
-        if (block.timestamp <= startTime_) return 0;
-
-        uint256 stopTime_ = stopTime();
-        if (block.timestamp < stopTime_) return block.timestamp - startTime_;
-
-        return stopTime_ - startTime_;
-    }
-
-    /**
-     * @notice Get this stream's token balance vs the token amount required to meet the commitment
-     * to recipient.
-     */
-    function tokenAndOutstandingBalance() public view returns (uint256, uint256) {
-        return (tokenBalance(), remainingBalance);
+        emit TokensRecovered(tokenAddress, amount);
     }
 
     /**
      * ░░░░░░░░░░░░░░░░░░░░░░░░░░░░░
-     *   INTERNAL FUNCTIONS
+     *   VIEW FUNCTIONS
      * ░░░░░░░░░░░░░░░░░░░░░░░░░░░░░
      */
 
     /**
      * @dev Helper function for `balanceOf` in calculating recipient's fair share of tokens, taking withdrawals into account.
+     * When a stream is cancelled the balance will be `recipientCancelBalance` until it's fully withdrawan, then
+     * it will remain at zero.
      */
-    function _recipientBalance() internal view returns (uint256) {
+    function recipientBalance() public view returns (uint256) {
         uint256 startTime_ = startTime();
         uint256 blockTime = block.timestamp;
 
         if (blockTime <= startTime_) return 0;
 
+        uint256 recipientCancelBalance_ = recipientCancelBalance;
+        if (recipientCancelBalance_ > 0) return recipientCancelBalance_;
```

Comment:
I think this introduces a problem.
with this change, after calling `cancel`, `withdraw` will work because it calls `recipientBalance()`, and because of the unchecked block it will succeed and also underflow `remainingBalance`.

I think `recipientBalance()` should not return `recipientCancelBalance` , i.e it should return 0 for cancelled streams

Reply:
Absolutely, I wanted to remove this and forgot!

Reply:
it's gone